app/controllers/admin/deploy_groups_controller.rb

```diff
@@ -52,46 +53,47 @@ def destroy
   end
 
   def deploy_all
-    @stages = Project.all.flat_map do |project|
-      environment = Environment.find(params[:environment_id]) if params[:environment_id]
+    environment = deploy_group.environment
+    deploys = deploy_group.stages.map do |stage|
+      template_stage = environment.template_stages.where(project: stage.project).first
+      next unless template_stage
 
-      stages = stages_in_same_environment(project, environment)
-      next unless deploy = stages.map(&:last_successful_deploy).compact.sort_by(&:created_at).last
-      stages.map { |s| [s, deploy] }
-    end.compact
-  end
+      last_success_deploy = template_stage.last_successful_deploy
+      next unless last_success_deploy
 
-  def deploy_all_now
-    deploys = params.require(:stages).map do |stage|
-      stage_id, reference = stage.split("-", 2)
-      stage = Stage.find(stage_id)
-      stage = new_stage_with_group(stage) unless only_to_current_group?(stage)
       deploy_service = DeployService.new(current_user)
-      deploy_service.deploy!(stage, reference: reference)
-    end
-    redirect_to deploys_path(ids: deploys.map(&:id))
-  end
-
-  private
+      deploy_service.deploy!(stage, reference: last_success_deploy.reference)
+    end.compact
 
-  def only_to_current_group?(stage)
-    stage.deploy_groups.map(&:id) == [deploy_group.id]
+    if deploys.empty?
+      flash[:error] = "There were no stages ready for deploy."
+      redirect_to deploys_path
+    else
+      redirect_to deploys_path(ids: deploys.map(&:id))
+    end
   end
 
-  def stages_in_same_environment(project, environment)
-    environment ||= deploy_group.environment
-    stages = project.stages.select do |stage|
-      stage.script.include?("$DEPLOY_GROUPS") && # is dynamic
-        stage.deploy_groups.where(environment: environment).exists? # is made to go to this environment
+  def create_all_stages
+    # No more than one stage, per project, per deploy_group
+    # Note: you can call this multiple times, and it will create missing stages, but no redundant stages.
+    environment = deploy_group.environment
+    Project.where(include_new_deploy_groups: true).each do |project|
```

Comment:
slightly misleading since the deploy group does not need to be new to do a the deploy-all ...